### DISCUSSION TITLE
Recursive interate and skip some tags

### DISCUSSION THREAD

**Discussion by filipenanclarez** [renggli/dart-xml#154] [Q&A] [answered]
I need interate over xml, but, every time i found an specific tag, i need skip all this tag and this tag childs.

I try to use `descendents` helper, but after skip the tag, this still interate over this tags childs.

How can i do this?

**Comment by renggli** [comment, marked as answer]
Probably the easiest is to verify that none of the ancestors of your matches also satisfy the condition:

```dart
final nodes = document.descendants
    // Find all the nodes that satisfy the condition.
    .where((node) => predicate(node))
    // Exclude the nodes that have parents satisfying the condition.
    .where((node) => !node.ancestors.any(predicate));
```

If you are only interested in elements, you might want to replace `descendants` and `ancestors` with `descendantElements` and `ancestorElements`. Also you might want to combine the `where` clauses into a single expression.

**Comment by renggli** [comment]
A more manual way would be to traverse the tree recursively:

```dart
List<XmlNode> find(XmlNode node, bool Function(XmlNode) predicate) {
  if (predicate(node)) {
    // Return a matching node, ...
    return [node];
  } else {
    // otherwise recurse into the children.
    return [
      ...node.attributes.expand((child) => find(child, predicate)),
      ...node.children.expand((child) => find(child, predicate)),
    ];
  }
}
```

This code could be further transformed into an iterative solution, and eventually into a generally reusable `Iterator` (which could be useful to include in this library).
